File: examples/java-tool-agent/src/test/java/example/AgentTest.java
package example;

import com.google.gson.*;
import com.sun.net.httpserver.HttpServer;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class AgentTest {
    @Test void wordCount() {
        assertEquals("{\"words\":4}", Agent.invoke("count_words", "{\"text\":\"Hello  世界\\u3000Agent\\n365\"}"));
        assertEquals("{\"words\":0}", Agent.invoke("count_words", "{\"text\":\"\"}"));
        assertEquals("{\"text\":\"B😀A\"}", Agent.invoke("reverse_text", "{\"text\":\"A😀B\"}"));
        assertTrue(Agent.invoke("count_words", "{}").contains("error"));
        assertTrue(Agent.invoke("unknown", "{\"text\":\"x\"}").contains("unknown tool"));
    }
    @Test void offlineAgent() throws Exception {
        var model = new Agent.OfflineModel("Hello Agent 365");
        assertEquals("Word count: 3", Agent.run(model, model.text, 6));
        assertEquals(2, model.requests);
        assertThrows(Exception.class, () -> Agent.run(new Agent.OfflineModel("x"), "x", 1));
    }
    @Test void providerIsolation() {
        assertThrows(IllegalArgumentException.class, () -> Agent.Provider.from(Map.of("MODEL_PROVIDER", "openai", "MISTRAL_API_KEY", "wrong-provider")));
        var provider = Agent.Provider.from(Map.of("MODEL_PROVIDER", "gemini", "GEMINI_API_KEY", "offline-key"));
        assertEquals("generativelanguage.googleapis.com", provider.endpoint().getHost());
        assertFalse(provider.toString().contains("offline-key"));
    }
    @Test void httpToolLoop() throws Exception {
        var calls = new AtomicInteger();
        var server = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
        server.createContext("/chat", exchange -> {
            var request = JsonParser.parseString(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8)).getAsJsonObject();
            assertEquals("Bearer offline-key", exchange.getRequestHeaders().getFirst("Authorization"));
            assertEquals(2, request.getAsJsonArray("tools").size());
            JsonObject message;
            if (calls.incrementAndGet() == 1) message = Agent.object(Map.of("role", "assistant", "tool_calls", java.util.List.of(Map.of(
                "id", "http-count", "type", "function", "function", Map.of("name", "count_words", "arguments", "{\"text\":\"Hello Agent 365\"}")))));
            else {
                var messages = request.getAsJsonArray("messages");
                assertEquals("{\"words\":3}", messages.get(messages.size()-1).getAsJsonObject().get("content").getAsString());
                message = Agent.object(Map.of("role", "assistant", "content", "Word count: 3"));
            }
            byte[] bytes = Agent.JSON.toJson(Map.of("choices", java.util.List.of(Map.of("message", message)))).getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, bytes.length);
            try (var body = exchange.getResponseBody()) { body.write(bytes); }
        });
        server.start();
        try (var model = new Agent.HttpModel(new Agent.Provider("fixture", "offline-model", "offline-key",
            URI.create("http://localhost:" + server.getAddress().getPort() + "/chat")))) {
            assertEquals("Word count: 3", Agent.run(model, "Count words", 6));
            assertEquals(2, calls.get());
        } finally { server.stop(0); }
    }
    @Test void responseLimit() {
        var body = new Agent.LimitedBody();
        var canceled = new java.util.concurrent.atomic.AtomicBoolean();
        body.onSubscribe(new java.util.concurrent.Flow.Subscription() {
            public void request(long count) {}
            public void cancel() { canceled.set(true); }
        });
        body.onNext(java.util.List.of(java.nio.ByteBuffer.allocate(Agent.MAX_RESPONSE_BYTES + 1)));
        assertTrue(canceled.get());
        assertTrue(body.getBody().toCompletableFuture().isCompletedExceptionally());
    }
}

File: examples/go-tool-agent/agent_test.go
package main

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestWordCount(t *testing.T) {
	if got := invokeTool("count_words", `{"text":"Hello  世界\u3000Agent\n365"}`); got != `{"words":4}` {
		t.Fatal(got)
	}
	if got := invokeTool("count_words", `{"text":""}`); got != `{"words":0}` {
		t.Fatal(got)
	}
	if !strings.Contains(invokeTool("count_words", `{}`), "error") {
		t.Fatal("missing text accepted")
	}
	if !strings.Contains(invokeTool("unknown", `{"text":"x"}`), "unknown tool") {
		t.Fatal("unknown tool accepted")
	}
	if got := invokeTool("reverse_text", `{"text":"A😀B"}`); got != `{"text":"B😀A"}` {
		t.Fatal(got)
	}
}

func TestOfflineAgent(t *testing.T) {
	model := &OfflineModel{Text: "Hello Agent 365"}
	output, err := runAgent(context.Background(), model, model.Text, 6)
	if err != nil || output != "Word count: 3" || model.Requests != 2 {
		t.Fatalf("%q %v", output, err)
	}
	_, err = runAgent(context.Background(), &OfflineModel{Text: "x"}, "x", 1)
	if err == nil {
		t.Fatal("turn limit was ignored")
	}
}

func TestHTTPToolLoop(t *testing.T) {
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		if r.Header.Get("Authorization") != "Bearer offline-key" {
			t.Error("wrong auth header")
		}
		var body map[string]any
		if json.NewDecoder(r.Body).Decode(&body) != nil {
			t.Fatal("invalid request")
		}
		if body["model"] != "offline-model" || len(body["tools"].([]any)) != 2 {
			t.Error("lost model/tools")
		}
		w.Header().Set("Content-Type", "application/json")
		if calls == 1 {
			_, _ = w.Write([]byte(`{"choices":[{"message":{"role":"assistant","tool_calls":[{"id":"http-count","type":"function","function":{"name":"count_words","arguments":"{\"text\":\"Hello Agent 365\"}"}}]}}]}`))
		} else {
			messages := body["messages"].([]any)
			result := messages[len(messages)-1].(map[string]any)
			if result["tool_call_id"] != "http-count" || result["content"] != `{"words":3}` {
				t.Error("lost tool result")
			}
			_, _ = w.Write([]byte(`{"choices":[{"message":{"role":"assistant","content":"Word count: 3"}}]}`))
		}
	}))
	defer server.Close()
	model := &HTTPModel{Endpoint: server.URL, Key: "offline-key", Name: "offline-model", Client: server.Client()}
	output, err := runAgent(context.Background(), model, "Count three words", 6)
	if err != nil || output != "Word count: 3" || calls != 2 {
		t.Fatalf("%q %v", output, err)
	}
}

func TestProviderIsolationAndHTTPFailure(t *testing.T) {
	values := map[string]string{"MODEL_PROVIDER": "openai", "MISTRAL_API_KEY": "must-not-be-reused"}
	if _, err := providerFromEnv(func(k string) string { return values[k] }); err == nil {
		t.Fatal("cross-provider key reuse")
	}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(403) }))
	defer server.Close()
	_, err := (&HTTPModel{Endpoint: server.URL, Key: "offline-key", Client: server.Client()}).Complete(context.Background(), nil)
	if err == nil || !strings.Contains(err.Error(), "HTTP 403") || strings.Contains(err.Error(), "offline-key") {
		t.Fatal(err)
	}
}

func TestResponseLimit(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		_, _ = w.Write([]byte(strings.Repeat("x", maxResponseBytes+1)))
	}))
	defer server.Close()
	_, err := (&HTTPModel{Endpoint: server.URL, Key: "offline-key", Client: server.Client()}).Complete(context.Background(), nil)
	if err == nil || !strings.Contains(err.Error(), "2 MiB") {
		t.Fatal(err)
	}
}
